juju/juju
# [JUJU-1368] Move more apis off the client facade

Move more methods off the client facade.
The ModelUserInfo api is moved to the UserManager facade. This api needs the model uuid to gather users for that model. The UserManager facade is a controller facade so the model uuid needs to be passed in.

There were also methods to stream / download charms and blobs (in this case agent binaries). These methods are used by the migration worker, but they don't use a versioned api facade but rather perform an http GET against the underlying http connection. So we consolidate existing http code and move the generic blob streaming to `api/http` and move the charm download wrapper to the `api/charms` package.

Finally, the AgentVersion behaviour was wrong (because it reported the version of the running agent and not the model). It is replaced with a call to extract the model version from the model config. This is only used deploying local charms.

When moving the code, existing JujuConnSuite tests, where feasible, we rewritten using gomock.

## QA steps

bootstrap a controller, add a model, deploy some charms
juju users mymodel
juju users
check the above commands with an older controller

bootstrap another controller and migrate the model to it
after migration, add another unit to the model
